### PR TITLE
feat: implement success cleanup for old swapped tiles on update

### PR DESCRIPTION
**Pull request by ronenkapelian**
On success oriented scenario - discrete cleanup should clean old tiles on updating swap mode (previous display path)

| Question                | Answer                                                                          |
| ---------------- | -------------------------------------------------------------------------- |
| Bug fix         | ✔                                                                        |
| New feature     | ✔                                                                        |
| Breaking change | ✖                                                                        |
| Deprecations    | ✖                                                                        |
| Documentation   | ✖                                                                        |
| Tests added     | ✔                                                                        |
| Chore            | ✖                                                                       |

1. a.scan running jobs by swap type.
b. validate no running exporting jobs on previous version. if running - skip, if no - execute cleanup
c. clean old display path tiles and source directory of raw data.
2. support s3 & fs providers

3. fixed bug - crash when FS try remove not exisitng directory.

### CODE REVIEW

File: config/custom-environment-variables.json [hunk 1]
```diff
@@ -84,7 +84,9 @@
   },
   "new_ingestion_job_type": "NEW_INGESTION_JOB_TYPE",
   "update_ingestion_job_type": "UPDATE_INGESTION_JOB_TYPE",
+  "swap_update_ingestion_job_type": "SWAP_UPDATE_INGESTION_JOB_TYPE",
```
Comment (asafMasa) [inline config/custom-environment-variables.json]: Move all job types to sub-group for readability

File: src/cleanupCommand/cleanupManager.ts [hunk 2]
```diff
@@ -49,14 +53,29 @@ export class CleanupManager {
...
       const currentBatch = notCleanedAndSuccess.slice(i, i + this.discreteBatchSize);
       const blackListFlitteredBatch = this.sourceBlackList.length > 0 ? this.filterBlackListSourcesTasks(currentBatch) : currentBatch;
-      await this.sourcesProvider.deleteDiscretes(blackListFlitteredBatch);
+      await this.sourcesProvider.deleteDiscretes(blackListFlitteredBatch, false);
       await this.jobManager.markAsCompletedAndRemoveFiles(currentBatch);
     }
   }
 
+  public async cleanSuccessfulSwappedLayersTasks(ingestionJobType: string): Promise<void> {
+    const notCleanedAndSuccess = await this.jobManager.getSuccessNotCleanedIngestionJobs(ingestionJobType);
```
Comment (asafMasa) [inline src/cleanupCommand/cleanupManager.ts]: add info log of "Starting Clean Successful Swapped Layers Logic"

File: src/cleanupCommand/cleanupManager.ts [hunk 3]
```diff
@@ -49,14 +53,29 @@ export class CleanupManager {
...
+    for (let i = 0; i < notCleanedAndSuccess.length; i += this.discreteBatchSize) {
```
Comment (asafMasa) [inline src/cleanupCommand/cleanupManager.ts:111]: rename method of "getSuccessNotCleanedIngestionJobs" to "getSuccessNotCleanedJobs"

File: src/cleanupCommand/cleanupManager.ts [hunk 4]
```diff
@@ -49,14 +53,29 @@ export class CleanupManager {
...
     for (let i = 0; i < notCleanedAndSuccess.length; i += this.discreteBatchSize) {
       const currentBatch = notCleanedAndSuccess.slice(i, i + this.discreteBatchSize);
       const blackListFlitteredBatch = this.sourceBlackList.length > 0 ? this.filterBlackListSourcesTasks(currentBatch) : currentBatch;
-      await this.sourcesProvider.deleteDiscretes(blackListFlitteredBatch);
+      await this.sourcesProvider.deleteDiscretes(blackListFlitteredBatch, false);
       await this.jobManager.markAsCompletedAndRemoveFiles(currentBatch);
     }
   }
 
+  public async cleanSuccessfulSwappedLayersTasks(ingestionJobType: string): Promise<void> {
```
Comment (asafMasa) [inline src/cleanupCommand/cleanupManager.ts]: rename "ingestionJobType" variable to "swappJobType"

File: src/cleanupCommand/cleanupManager.ts [hunk 5]
```diff
@@ -1,16 +1,18 @@
-import { IngestionParams } from '@map-colonies/mc-model-types';
+// import { IngestionParams } from '@map-colonies/mc-model-types';
```
Comment (asafMasa) [inline src/cleanupCommand/cleanupManager.ts]: delete comment

File: config/custom-environment-variables.json [hunk 6]
```diff
@@ -30,7 +30,10 @@
   "fs": {
     "tiles_location": "FS_TILES_LOCATION",
     "sources_location": "FS_SOURCES_LOCATION",
-    "blacklist_sources_location": "FS_BLACK_LIST_SOURCES_LOCATION"
+    "blacklist_sources_location": {
```
Comment (asafMasa) [inline config/custom-environment-variables.json:33]: fix prettier

File: src/cleanupCommand/cleanupManager.ts [hunk 7]
```diff
@@ -33,7 +37,8 @@ export class CleanupManager {
     for (let i = 0; i < notCleanedAndFailedNew.length; i += this.discreteBatchSize) {
       const currentBatch = notCleanedAndFailedNew.slice(i, i + this.discreteBatchSize);
       const expiredBatch = await this.deleteExpiredFailedTasksSources(currentBatch);
-      await this.tileProvider.deleteDiscretes(currentBatch);
+      const tilesDirectories = this.getCurrentTilesLocation(currentBatch);
```
Comment (asafMasa) [inline src/cleanupCommand/cleanupManager.ts]: Please add log.Info when major methods are called (with job types), such as:
* cleanFailedIngestionTasks
* cleanSuccessfulIngestionTasks
* cleanSuccessfulSwappedLayersTasks
* cleanFailedIncomingSyncTasks

File: src/storageProviders/s3/s3TileStorageProvider.ts [hunk 8]
```diff
@@ -41,8 +40,8 @@ export class S3TileStorageProvider implements IStorageProvider {
     this.batchSize = config.get<number>('batch_size.tilesDeletion');
   }
 
-  public async deleteDiscretes(discreteArray: IJob<IngestionParams>[]): Promise<void> {
-    const s3PreFixes = this.parseLocation(discreteArray);
+  public async deleteDiscretes(discreteLocationArray: IDataLocation[]): Promise<void> {
+    const s3PreFixes = this.concatDirectories(discreteLocationArray);
```
Comment (asafMasa) [inline src/storageProviders/s3/s3TileStorageProvider.ts:44]: please add log.info of "deleting tiles in location {...}"

File: tests/testData.ts [hunk 9]
```diff
@@ -1,5 +1,5 @@
-import { IngestionParams } from '@map-colonies/mc-model-types';
-import { IJob } from '../src/common/interfaces';
+// import { IngestionParams } from '@map-colonies/mc-model-types';
```
Comment (asafMasa) [inline tests/testData.ts]: remove this line

File: src/clients/mapproxyClient.ts [hunk 10]
```diff
@@ -1,17 +1,17 @@
 import { inject, singleton } from 'tsyringe';
 import { Logger } from '@map-colonies/js-logger';
 import { HttpClient, IHttpRetryConfig } from '@map-colonies/mc-utils';
-import { IngestionParams } from '@map-colonies/mc-model-types';
+// import { IngestionParams } from '@map-colonies/mc-model-types';
```
Comment (asafMasa) [inline src/clients/mapproxyClient.ts]: remove line

File: src/common/interfaces.ts [hunk 11]
```diff
@@ -20,3 +21,18 @@ export interface IJob<T> {
...
+  previousProductVersion: string;
+}
+
+export interface IDataLocation {
+  directory: string;
+}
+export interface ITilesLocation extends IDataLocation {
+  subDirectory: string;
+}
+// TODO - Temporary implementation till it will be implemented inside mc-models on IngestionParams interface.
```
Comment (asafMasa) [inline src/common/interfaces.ts]: this shouldn't be in mc-models so it isn't temporery...

File: src/storageProviders/fs/fsStorageProviderBase.ts [hunk 12]
```diff
@@ -1,21 +1,23 @@
-import { promises } from 'fs';
+import { promises, existsSync } from 'fs';
```
Comment (asafMasa) [inline src/storageProviders/fs/fsStorageProviderBase.ts]: please change the import in all app to: "from 'node:fs'"

File: src/storageProviders/fs/fsStorageProviderBase.ts [hunk 13]
```diff
@@ -1,21 +1,23 @@
...
-    const directories = this.parseLocation(discreteArray);
+  public async deleteDiscretes(discreteLocationArray: IDataLocation[]): Promise<void> {
+    const directories = this.concatDirectories(discreteLocationArray);
     let batchArray = [];
     for (let i = 0; i < directories.length; i += this.batchSize) {
       batchArray = directories.slice(i, i + this.batchSize);
       this.logger.info(`Deleting directories from FS in path: [${batchArray.join(',')}]`);
-      const promiseDeleteArray = batchArray.map(async (directory: string) => promises.rmdir(directory, { recursive: true }));
+      const promiseDeleteArray = batchArray.map(async (directory: string) =>
+        existsSync(directory) ? promises.rmdir(directory, { recursive: true }) : this.logger.warn(`Current directory: ${directory} is not exists!`)
```
Comment (asafMasa) [inline src/storageProviders/fs/fsStorageProviderBase.ts]: why do you use existsSync? try https://dev.to/hunterkohler/comment/1gkn0 approach such as [fs.Stats](https://iojs.org/api/fs.html#fs_fs_stat_path_callback)

File: src/cleanupCommand/cleanupManager.ts [hunk 14]
```diff
@@ -96,14 +128,64 @@ export class CleanupManager {
...
   }
+
+  private getSourcesLocation(discreteArray: IJob<IWithCleanDataIngestionParams>[]): IDataLocation[] {
+    const sourcesDirectories: IDataLocation[] = discreteArray.map((discrete) => {
+      return { directory: discrete.parameters.originDirectory };
+    });
+    return sourcesDirectories;
+  }
+
+  private getCurrentTilesLocation(discreteArray: IJob<IWithCleanDataIngestionParams>[]): ITilesLocation[] {
```
Comment (asafMasa) [inline src/cleanupCommand/cleanupManager.ts]: rename to "getTilesLocation"

File: src/storageProviders/fs/fsStorageProviderBase.ts [hunk 15]
```diff
@@ -1,21 +1,23 @@
...
 import { IStorageProvider } from '../iStorageProvider';
+import { IDataLocation } from '../../common/interfaces';
 
 export abstract class FsStorageProviderBase implements IStorageProvider {
   public constructor(protected batchSize: number, protected logger: Logger) {}
 
-  public async deleteDiscretes(discreteArray: IJob<IngestionParams>[]): Promise<void> {
-    const directories = this.parseLocation(discreteArray);
+  public async deleteDiscretes(discreteLocationArray: IDataLocation[]): Promise<void> {
+    const directories = this.concatDirectories(discreteLocationArray);
```
Comment (asafMasa) [inline src/storageProviders/fs/fsStorageProviderBase.ts:10]: add info log of deleted paths

File: src/cleanupCommand/cleanupManager.ts [hunk 16]
```diff
@@ -49,14 +54,40 @@ export class CleanupManager {
...
 
+  public async cleanSuccessfulSwappedLayersTasks(ingestionJobType: string): Promise<void> {
+    const notCleanedAndSuccess = await this.jobManager.getSuccessNotCleanedIngestionJobs(ingestionJobType);
+    for (let i = 0; i < notCleanedAndSuccess.length; i += this.discreteBatchSize) {
+      const currentBatch = notCleanedAndSuccess.slice(i, i + this.discreteBatchSize);
+      console.log(currentBatch,'batch')
+      // cleaning tiles of all success jobs excluding layer that been exporting on current iteration.
+      const notRunningExportFilteredBatch = await this.filterFromRunningExportJobs(currentBatch);
+      const tilesDirectories = this.getSwappedTilesLocation(notRunningExportFilteredBatch);
+      console.log(tilesDirectories,'*************************')
```
Comment (asafMasa) [inline src/cleanupCommand/cleanupManager.ts]: console log???

File: src/cleanupCommand/cleanupManager.ts [hunk 17]
```diff
@@ -49,14 +54,40 @@ export class CleanupManager {
...
+      await this.tileProvider.deleteDiscretes(tilesDirectories);
```
Comment (asafMasa) [inline src/cleanupCommand/cleanupManager.ts:122]: find all "deleteDiscretes" and add info log of deleted paths

File: src/cleanupCommand/cleanupManager.ts [hunk 18]
```diff
@@ -49,14 +54,40 @@ export class CleanupManager {
 
   public async cleanSuccessfulIngestionTasks(ingestionJobType: string): Promise<void> {
     const notCleanedAndSuccess = await this.jobManager.getSuccessNotCleanedIngestionJobs(ingestionJobType);
+
     for (let i = 0; i < notCleanedAndSuccess.length; i += this.discreteBatchSize) {
       const currentBatch = notCleanedAndSuccess.slice(i, i + this.discreteBatchSize);
       const blackListFlitteredBatch = this.sourceBlackList.length > 0 ? this.filterBlackListSourcesTasks(currentBatch) : currentBatch;
-      await this.sourcesProvider.deleteDiscretes(blackListFlitteredBatch);
+      const sourcesDirectories = this.getSourcesLocation(blackListFlitteredBatch);
+      await this.sourcesProvider.deleteDiscretes(sourcesDirectories);
```
Comment (asafMasa) [inline src/cleanupCommand/cleanupManager.ts:98]: add log inside sourcesProvider.deleteDiscretes method of path that needs to delete

File: src/cleanupCommand/cleanupManager.ts [hunk 19]
```diff
@@ -49,14 +54,40 @@ export class CleanupManager {
...
       await this.jobManager.markAsCompletedAndRemoveFiles(currentBatch);
     }
   }
 
+  public async cleanSuccessfulSwappedLayersTasks(ingestionJobType: string): Promise<void> {
+    const notCleanedAndSuccess = await this.jobManager.getSuccessNotCleanedIngestionJobs(ingestionJobType);
+    for (let i = 0; i < notCleanedAndSuccess.length; i += this.discreteBatchSize) {
+      const currentBatch = notCleanedAndSuccess.slice(i, i + this.discreteBatchSize);
+      console.log(currentBatch,'batch')
```
Comment (asafMasa) [inline src/cleanupCommand/cleanupManager.ts]: console log???

File: src/cleanupCommand/cleanupManager.ts [hunk 20]
```diff
@@ -49,14 +54,40 @@ export class CleanupManager {
...
+
+      // clean source data only for jobs excluded the blacklist
+      const blackListFlitteredBatch =
+        this.sourceBlackList.length > 0 ? this.filterBlackListSourcesTasks(notRunningExportFilteredBatch) : notRunningExportFilteredBatch;
+      const sourcesDirectories = this.getSourcesLocation(blackListFlitteredBatch);
+      console.log(blackListFlitteredBatch, blackListFlitteredBatch,this.sourceBlackList )
+      await this.sourcesProvider.deleteDiscretes(sourcesDirectories);
+
+      if (notRunningExportFilteredBatch.length) {
+        await this.jobManager.markAsCompletedAndRemoveFiles(notRunningExportFilteredBatch);
```
Comment (asafMasa) [inline src/cleanupCommand/cleanupManager.ts:143]: log what jobs are marked as completed, and which are skipped due to which blacklist

File: src/cleanupCommand/cleanupManager.ts [hunk 21]
```diff
@@ -49,14 +54,40 @@ export class CleanupManager {
...
+      const tilesDirectories = this.getSwappedTilesLocation(notRunningExportFilteredBatch);
+      console.log(tilesDirectories,'*************************')
+      await this.tileProvider.deleteDiscretes(tilesDirectories);
+
+      // clean source data only for jobs excluded the blacklist
+      const blackListFlitteredBatch =
+        this.sourceBlackList.length > 0 ? this.filterBlackListSourcesTasks(notRunningExportFilteredBatch) : notRunningExportFilteredBatch;
+      const sourcesDirectories = this.getSourcesLocation(blackListFlitteredBatch);
+      console.log(blackListFlitteredBatch, blackListFlitteredBatch,this.sourceBlackList )
+      await this.sourcesProvider.deleteDiscretes(sourcesDirectories);
```
Comment (asafMasa) [inline src/cleanupCommand/cleanupManager.ts:136]: log what jobs sources are beeing deleted, and which are skipped due to blacklist

File: src/cleanupCommand/cleanupManager.ts [hunk 22]
```diff
@@ -96,14 +128,64 @@ export class CleanupManager {
     return filteredTasks;
   }
 
-  private async deleteExpiredFailedTasksSources(tasks: IJob<IngestionParams>[]): Promise<IJob<IngestionParams>[]> {
+  private async filterFromRunningExportJobs(jobs: IJob<IWithCleanDataIngestionParams>[]): Promise<IJob<IWithCleanDataIngestionParams>[]> {
+    const runningExportJobs = await this.jobManager.getInProgressJobs(this.exporterJobType);
```
Comment (asafMasa) [inline src/cleanupCommand/cleanupManager.ts:198]: why don't you add more criterias to search such as catalogId?

File: src/cleanupCommand/cleanupManager.ts [hunk 23]
```diff
@@ -96,14 +128,64 @@ export class CleanupManager {
...
+        (job) => cleanupData?.previousProductVersion === job.version && swapJob.parameters.metadata.productId === job.resourceId
+      );
+      if (!filteredJobs.length) {
+        cleanupReadyJobs.push(swapJob);
+      }
+    }
+    return cleanupReadyJobs;
+  }
+
+  private async deleteExpiredFailedTasksSources(tasks: IJob<IWithCleanDataIngestionParams>[]): Promise<IJob<IWithCleanDataIngestionParams>[]> {
```
Comment (asafMasa) [inline src/cleanupCommand/cleanupManager.ts:216]: add info log for deleted expired sources

File: src/clients/jobManagerClient.ts [hunk 24]
```diff
@@ -1,9 +1,9 @@
 import { inject, singleton } from 'tsyringe';
 import { Logger } from '@map-colonies/js-logger';
 import { HttpClient, IHttpRetryConfig } from '@map-colonies/mc-utils';
-import { IngestionParams } from '@map-colonies/mc-model-types';
+// import { IngestionParams } from '@map-colonies/mc-model-types';
```
Comment (asafMasa) [inline src/clients/jobManagerClient.ts]: remove

File: src/clients/jobManagerClient.ts [hunk 25]
```diff
@@ -14,32 +14,41 @@ export class JobManagerClient extends HttpClient {
...
       isCleaned: false,
       status: 'Failed',
       type: jobType,
       shouldReturnTasks: false,
     });
   }
 
-  public async getFailedAndNotCleanedIncomingSyncJobs(): Promise<IJob<IngestionParams>[]> {
-    const failed = await this.get<IJob<IngestionParams>[]>('/jobs', {
+  public async getInProgressJobs(jobType: string): Promise<IJob<IWithCleanDataIngestionParams>[]> {
```
Comment (asafMasa) [inline src/clients/jobManagerClient.ts:25]: think abourt uniting the get jobs methods to obe base that is being used with different parms

**Review by asafMasa — CHANGES_REQUESTED** [review]
Good work!

File: config/custom-environment-variables.json (hunk 6 above)
Comment (ronenkapelian) [reply config/custom-environment-variables.json:33]: done

File: config/custom-environment-variables.json (hunk 1 above)
Comment (ronenkapelian) [reply config/custom-environment-variables.json]: done

File: src/clients/jobManagerClient.ts (hunk 24 above)
Comment (ronenkapelian) [reply src/clients/jobManagerClient.ts]: done

File: src/cleanupCommand/cleanupManager.ts (hunk 5 above)
Comment (ronenkapelian) [reply src/cleanupCommand/cleanupManager.ts]: done

File: src/clients/mapproxyClient.ts (hunk 10 above)
Comment (ronenkapelian) [reply src/clients/mapproxyClient.ts]: done

Comment (ronenkapelian) [reply src/clients/mapproxyClient.ts]: done

Comment (ronenkapelian) [reply src/clients/mapproxyClient.ts]: done

File: src/cleanupCommand/cleanupManager.ts (hunk 7 above)
Comment (ronenkapelian) [reply src/cleanupCommand/cleanupManager.ts]: done

File: src/cleanupCommand/cleanupManager.ts (hunk 4 above)
Comment (ronenkapelian) [reply src/cleanupCommand/cleanupManager.ts]: done

File: src/cleanupCommand/cleanupManager.ts (hunk 18 above)
Comment (ronenkapelian) [reply src/cleanupCommand/cleanupManager.ts:98]: done

File: src/cleanupCommand/cleanupManager.ts (hunk 3 above)
Comment (ronenkapelian) [reply src/cleanupCommand/cleanupManager.ts:111]: done

File: src/cleanupCommand/cleanupManager.ts (hunk 17 above)
Comment (ronenkapelian) [reply src/cleanupCommand/cleanupManager.ts:122]: done

File: src/cleanupCommand/cleanupManager.ts (hunk 21 above)
Comment (ronenkapelian) [reply src/cleanupCommand/cleanupManager.ts:136]: done

File: src/cleanupCommand/cleanupManager.ts (hunk 20 above)
Comment (ronenkapelian) [reply src/cleanupCommand/cleanupManager.ts:143]: done

File: src/cleanupCommand/cleanupManager.ts (hunk 14 above)
Comment (ronenkapelian) [reply src/cleanupCommand/cleanupManager.ts]: done

File: src/clients/mapproxyClient.ts (hunk 10 above)
Comment (ronenkapelian) [reply src/clients/mapproxyClient.ts]: done

File: src/common/interfaces.ts (hunk 11 above)
Comment (ronenkapelian) [reply src/common/interfaces.ts]: done

File: src/storageProviders/fs/fsStorageProviderBase.ts (hunk 12 above)
Comment (ronenkapelian) [reply src/storageProviders/fs/fsStorageProviderBase.ts]: done

File: src/storageProviders/fs/fsStorageProviderBase.ts (hunk 13 above)
Comment (ronenkapelian) [reply src/storageProviders/fs/fsStorageProviderBase.ts]: I'd refactored the mechanism and avoid using existsSync

File: src/storageProviders/fs/fsStorageProviderBase.ts (hunk 15 above)
Comment (ronenkapelian) [reply src/storageProviders/fs/fsStorageProviderBase.ts:10]: done

File: src/storageProviders/s3/s3TileStorageProvider.ts (hunk 8 above)
Comment (ronenkapelian) [reply src/storageProviders/s3/s3TileStorageProvider.ts:44]: done

File: tests/testData.ts (hunk 9 above)
Comment (ronenkapelian) [reply tests/testData.ts]: done

File: src/clients/jobManagerClient.ts (hunk 25 above)
Comment (ronenkapelian) [reply src/clients/jobManagerClient.ts:25]: after consulting with you, it was refactored as we decided, and issue will opened to future task, implementing the usage of mc-priority-queue pacakge.

File: src/cleanupCommand/cleanupManager.ts (hunk 2 above)
Comment (ronenkapelian) [reply src/cleanupCommand/cleanupManager.ts]: done

File: src/cleanupCommand/cleanupManager.ts [hunk 26]
```diff
@@ -1,81 +1,178 @@
...
     const notCleanedAndFailedUpdate = await this.jobManager.getFailedAndNotCleanedIngestionJobs(this.updateIngestionJobType);
     for (let i = 0; i < notCleanedAndFailedUpdate.length; i += this.discreteBatchSize) {
       const currentBatch = notCleanedAndFailedUpdate.slice(i, i + this.discreteBatchSize);
+      this.logger.info({
+        msg: `Will execute cleanup to ${currentBatch.length} failed jobs of type: '${this.updateIngestionJobType}'`,
+        batch: `${i + 1}/${Math.floor(notCleanedAndFailedUpdate.length / this.discreteBatchSize + 1)}`,
+      });
       const expiredBatch = await this.deleteExpiredFailedTasksSources(currentBatch);
       await this.jobManager.markAsCompletedAndRemoveFiles(expiredBatch);
+      this.logger.info({
```
Comment (asafMasa) [inline src/cleanupCommand/cleanupManager.ts:69]: why this log is different from the above log?

File: src/cleanupCommand/cleanupManager.ts [hunk 27]
```diff
@@ -1,81 +1,178 @@
...
   }
 
   public async cleanSuccessfulIngestionTasks(ingestionJobType: string): Promise<void> {
-    const notCleanedAndSuccess = await this.jobManager.getSuccessNotCleanedIngestionJobs(ingestionJobType);
+    this.logger.info({ msg: `Running Cleanup for success ingestion jobs of type: ${ingestionJobType}` });
+    const notCleanedAndSuccess = await this.jobManager.getSuccessNotCleanedJobs(ingestionJobType);
+
     for (let i = 0; i < notCleanedAndSuccess.length; i += this.discreteBatchSize) {
       const currentBatch = notCleanedAndSuccess.slice(i, i + this.discreteBatchSize);
       const blackListFlitteredBatch = this.sourceBlackList.length > 0 ? this.filterBlackListSourcesTasks(currentBatch) : currentBatch;
```
Comment (asafMasa) [inline src/cleanupCommand/cleanupManager.ts]: Can you please fix the spelling? (Flittered...)

File: src/cleanupCommand/cleanupManager.ts [hunk 28]
```diff
@@ -1,81 +1,178 @@
...
+      this.logger.info({
+        msg: `Will execute cleanup to ${notRunningExportFilteredBatch.length} success jobs of type: '${swappJobType}'`,
+        batch: `${i + 1}/${Math.floor(notCleanedAndSuccess.length / this.discreteBatchSize + 1)}`,
+      });
+      this.logger.debug({ jobIds: notRunningExportFilteredBatch.map((job) => job.id) });
+      const tilesDirectories = this.getSwappedTilesLocation(notRunningExportFilteredBatch);
+      await this.tileProvider.deleteDiscretes(tilesDirectories);
+
+      // clean source data only for jobs excluded the blacklist
+      const blackListFlitteredBatch =
```
Comment (asafMasa) [inline src/cleanupCommand/cleanupManager.ts]: Can you please fix the spelling? (Flittered...)

File: src/cleanupCommand/cleanupManager.ts [hunk 29]
```diff
@@ -85,8 +182,8 @@ export class CleanupManager {
     return filteredTasks;
   }
 
-  private filterBlackListSourcesTasks(tasks: IJob<IngestionParams>[]): IJob<IngestionParams>[] {
-    const filteredTasks: IJob<IngestionParams>[] = [];
+  private filterBlackListSourcesTasks(tasks: IJob<IWithCleanDataIngestionParams>[]): IJob<IWithCleanDataIngestionParams>[] {
+    const filteredTasks: IJob<IWithCleanDataIngestionParams>[] = [];
```
Comment (asafMasa) [inline src/cleanupCommand/cleanupManager.ts:187]: in the future consider using [filter](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter) method

File: src/common/interfaces.ts [hunk 30]
```diff
@@ -20,3 +22,17 @@ export interface IJob<T> {
...
+  previousProductVersion: string;
+}
+
+export interface IDataLocation {
+  directory: string;
+}
+export interface ITilesLocation extends IDataLocation {
+  subDirectory: string;
+}
+export interface IWithCleanDataIngestionParams extends IngestionParams {
```
Comment (asafMasa) [inline src/common/interfaces.ts:39]: add line separator

File: src/common/interfaces.ts [hunk 31]
```diff
@@ -20,3 +22,17 @@ export interface IJob<T> {
...
 }
+export interface ICleanupData {
+  previousRelativePath: string;
+  previousProductVersion: string;
+}
+
+export interface IDataLocation {
+  directory: string;
+}
+export interface ITilesLocation extends IDataLocation {
```
Comment (asafMasa) [inline src/common/interfaces.ts:35]: add line separator

File: src/common/interfaces.ts [hunk 32]
```diff
@@ -20,3 +22,17 @@ export interface IJob<T> {
   priority: number;
   tasks?: unknown[];
 }
+export interface ICleanupData {
```
Comment (asafMasa) [inline src/common/interfaces.ts:26]: add line separator

File: src/cleanupCommand/cleanupManager.ts [hunk 33]
```diff
@@ -1,81 +1,178 @@
...
     deleteDate.setDate(deleteDate.getDate() - FAILED_CLEANUP_DELAY);
+    this.logger.info({ msg: `Running Cleanup for failed incoming sync jobs of type: ${this.syncJobType}` });
     const notCleanedAndFailed = await this.jobManager.getFailedAndNotCleanedIncomingSyncJobs();
     for (let i = 0; i < notCleanedAndFailed.length; i += this.discreteBatchSize) {
       const currentBatch = notCleanedAndFailed.slice(i, i + this.discreteBatchSize);
+      this.logger.info({
+        msg: `Will execute cleanup to ${currentBatch.length} failed incoming sync jobs of type: '${this.syncJobType}'`,
+        batch: `${i + 1}/${Math.floor(notCleanedAndFailed.length / this.discreteBatchSize + 1)}`,
+      });
+      this.logger.debug({ jobIds: currentBatch.map((job) => job.id) });
```
Comment (asafMasa) [inline src/cleanupCommand/cleanupManager.ts]: please enter this debug log to the info log

File: src/cleanupCommand/cleanupManager.ts (hunk 23 above)
Comment (asafMasa) [reply src/cleanupCommand/cleanupManager.ts:216]: ?

File: src/cleanupCommand/cleanupManager.ts [hunk 34]
```diff
@@ -96,14 +193,67 @@ export class CleanupManager {
...
+        directory: discrete.parameters.metadata.id as string,
+        subDirectory: discrete.parameters.metadata.displayPath as string,
+      };
+    });
+    return tilesDirectories;
+  }
+
+  private getSwappedTilesLocation(discreteArray: IJob<IWithCleanDataIngestionParams>[]): ITilesLocation[] {
+    const tilesDirectories: ITilesLocation[] = discreteArray
+      .filter((v) => v.parameters.cleanupData)
```
Comment (asafMasa) [inline src/cleanupCommand/cleanupManager.ts]: rename "v" to appropriate object (job?)

File: src/cleanupCommand/cleanupManager.ts [hunk 35]
```diff
@@ -1,81 +1,178 @@
...
+    for (let i = 0; i < notCleanedAndSuccess.length; i += this.discreteBatchSize) {
+      const currentBatch = notCleanedAndSuccess.slice(i, i + this.discreteBatchSize);
+
+      // cleaning tiles of all success jobs excluding layer that been exporting on current iteration.
+      const notRunningExportFilteredBatch = await this.filterFromRunningExportJobs(currentBatch);
+      this.logger.info({
+        msg: `Will execute cleanup to ${notRunningExportFilteredBatch.length} success jobs of type: '${swappJobType}'`,
+        batch: `${i + 1}/${Math.floor(notCleanedAndSuccess.length / this.discreteBatchSize + 1)}`,
+      });
+      this.logger.debug({ jobIds: notRunningExportFilteredBatch.map((job) => job.id) });
```
Comment (asafMasa) [inline src/cleanupCommand/cleanupManager.ts]: please unite the debug log into the info log

File: tests/unit/storageProviders/fs/fsSourcesStorageProvider.spec.ts [hunk 36]
```diff
@@ -1,9 +1,9 @@
-import { promises } from 'fs';
+import fs from 'node:fs';
 import path from 'path';
```
Comment (asafMasa) [inline tests/unit/storageProviders/fs/fsSourcesStorageProvider.spec.ts]: please rename all "path" imports to "node:path"

**Review by asafMasa — CHANGES_REQUESTED** [review]
Great job!

File: tests/unit/storageProviders/fs/fsSourcesStorageProvider.spec.ts (hunk 36 above)
Comment (ronenkapelian) [reply tests/unit/storageProviders/fs/fsSourcesStorageProvider.spec.ts]: done

File: src/cleanupCommand/cleanupManager.ts (hunk 33 above)
Comment (ronenkapelian) [reply src/cleanupCommand/cleanupManager.ts]: done

File: src/cleanupCommand/cleanupManager.ts (hunk 26 above)
Comment (ronenkapelian) [reply src/cleanupCommand/cleanupManager.ts:69]: done

File: src/cleanupCommand/cleanupManager.ts (hunk 27 above)
Comment (ronenkapelian) [reply src/cleanupCommand/cleanupManager.ts]: done

File: src/cleanupCommand/cleanupManager.ts (hunk 35 above)
Comment (ronenkapelian) [reply src/cleanupCommand/cleanupManager.ts]: done

File: src/cleanupCommand/cleanupManager.ts (hunk 28 above)
Comment (ronenkapelian) [reply src/cleanupCommand/cleanupManager.ts]: done

File: src/cleanupCommand/cleanupManager.ts (hunk 33 above)
Comment (ronenkapelian) [reply src/cleanupCommand/cleanupManager.ts]: done

File: src/cleanupCommand/cleanupManager.ts (hunk 29 above)
Comment (ronenkapelian) [reply src/cleanupCommand/cleanupManager.ts:187]: ok

File: src/cleanupCommand/cleanupManager.ts (hunk 23 above)
Comment (ronenkapelian) [reply src/cleanupCommand/cleanupManager.ts:216]: done

File: src/cleanupCommand/cleanupManager.ts (hunk 34 above)
Comment (ronenkapelian) [reply src/cleanupCommand/cleanupManager.ts]: done

File: src/common/interfaces.ts (hunk 32 above)
Comment (ronenkapelian) [reply src/common/interfaces.ts:26]: done

File: src/common/interfaces.ts (hunk 31 above)
Comment (ronenkapelian) [reply src/common/interfaces.ts:35]: done

File: src/common/interfaces.ts (hunk 30 above)
Comment (ronenkapelian) [reply src/common/interfaces.ts:39]: done

File: tests/unit/storageProviders/fs/fsSourcesStorageProvider.spec.ts (hunk 36 above)
Comment (ronenkapelian) [reply tests/unit/storageProviders/fs/fsSourcesStorageProvider.spec.ts]: done

**Review by asafMasa — APPROVED** [review]

**Review by asafMasa — APPROVED** [review]

**Pull request merged**